stream-chat-android-core/src/main/java/io/getstream/chat/android/client/Result.kt

```diff
@@ -0,0 +1,212 @@
+     * Returns an original [Result] if the [Result] contains an error payload.
+     *
+     * @param f A lambda for mapping [Result] of [A] to [Result] of [C].
+     *
+     * @return A transformed instance of the [Result] or the original instance of the [Result].
+     */
+    @JvmSynthetic
+    public inline fun <C : Any> map(f: (A) -> C): Result<C, B> = flatMap { Success(f(it)) }
+
+    /**
+     * Returns a transformed [Result] of applying the given suspending [f] function if the [Result]
+     * contains a successful data payload.
+     * Returns an original [Result] if the [Result] contains an error payload.
+     *
+     * @param f A suspending lambda for mapping [Result] of [A] to [Result] of [C].
+     *
+     * @return A transformed instance of the [Result] or the original instance of the [Result].
+     */
+    @JvmSynthetic
+    public suspend inline fun <C : Any> mapSuspend(crossinline f: suspend (A) -> C): Result<C, B> =
+        flatMap { Success(f(it)) }
+
+    /**
+     * Returns a [Result] of [Unit] from any type of a [Result].
+     *
+     * @return [Result] of [Unit].
+     */
+    public fun toUnitResult(): Result<Unit, B> = map {}
+}
+
+/**
+ * Returns a transformed [Result] from results of the [f] if the [Result] contains a successful data payload.
+ * Returns an original [Result] if the [Result] contains an error payload.
+ *
+ * @param f A lambda that returns [Result] of [C].
+ *
+ * @return A transformed instance of the [Result] or the original instance of the [Result].
+ */
+@JvmSynthetic
+public inline fun <A : Any, B : ChatError, C : Any> Result<A, B>.flatMap(f: (A) -> Result<C, B>): Result<C, B> {
+    return when (this) {
+        is Result.Success -> f(this.value)
+        is Result.Failure -> this
+    }
+}
+
+/**
+ * Returns a transformed [Result] from results of the suspending [f] if the [Result] contains a successful data
+ * payload.
+ * Returns an original [Result] if the [Result] contains an error payload.
+ *
+ * @param f A suspending lambda that returns [Result] of [C].
+ *
+ * @return A transformed instance of the [Result] or the original instance of the [Result].
+ */
+@JvmSynthetic
+public suspend inline fun <A : Any, B : ChatError, C : Any> Result<A, B>.flatMapSuspend(
+    crossinline f: suspend (A) -> Result<C, B>,
+): Result<C, B> {
+    return when (this) {
+        is Result.Success -> f(this.value)
+        is Result.Failure -> this
+    }
+}
+
+/**
+ * Runs the [successSideEffect] lambda function if the [Result] contains a successful data payload.
+ *
+ * @param successSideEffect A lambda that receives the successful data payload.
+ *
+ * @return The original instance of the [Result].
+ */
+@JvmSynthetic
+public inline fun <A : Any, B : ChatError> Result<A, B>.onSuccess(
+    crossinline successSideEffect: (A) -> Unit,
+): Result<A, B> =
+    apply {
+        when (this) {
+            is Result.Success -> successSideEffect(this.value)
+            is Result.Failure -> Unit
+        }
+    }
```

Comment:
```suggestion
    also {
        when (it) {
            is Result.Success -> successSideEffect(it.value)
            is Result.Failure -> Unit
        }
    }
```

What about to use `also` instead of `apply`?
`apply` gives the impression we are "updading" the result before we return it